File: otel-android-mobile/src/main/java/io/opentelemetry/android/mobile/policy/RemoteGate.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.mobile.policy

import io.opentelemetry.android.mobile.instrumentation.Incubating
import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.atomic.AtomicReference

/**
 * Thread-safe single source of truth for the remote kill switch and global
 * head-sampling override delivered via the `sdk` block of the remote config.
 *
 * The gate is consulted on the hot path at the two telemetry emission choke
 * points — [io.opentelemetry.android.mobile.buffering.MobileLogRecordProcessor.onEmit]
 * for logs and [io.opentelemetry.android.mobile.sampling.DynamicSampler.shouldSample]
 * for spans. Both choke points share the **same** gate instance, which is wired
 * once at SDK init in [io.opentelemetry.android.mobile.MobileLoggerProvider].
 *
 * **State model.** State is held as a single immutable [GateState] behind one
 * `@Volatile` reference, so a hot-path read is a single volatile load with no
 * lock and no allocation. Writes (from config apply) publish a fresh immutable
 * snapshot, so a reader always sees an internally-consistent `(enabled, rate)`
 * pair — never a torn mix of an old `enabled` with a new `sampleRate`.
 *
 * **Fail-open.** A freshly-constructed gate is `enabled = true, sampleRate = 1.0`
 * (no restriction). A transient config-fetch failure must leave the prior state
 * untouched: callers simply do not call [apply] on failure. An applied config
 * whose `sdk` block is absent re-enables the SDK (absence == "no restriction"),
 * which is why [apply] maps a `null` [SdkConfig] back to the open defaults.
 *
 * @see SdkConfig for the parsed, clamped wire representation.
 */
@Incubating
class RemoteGate {

    /**
     * Immutable snapshot of the gate's two control values. Published atomically
     * via the enclosing [state] reference so readers never observe a torn pair.
     *
     * @property enabled `false` hard-drops all new telemetry at both choke points.
     * @property sampleRate global head-sampling fraction in `[0.0, 1.0]`.
     */
    data class GateState(
        val enabled: Boolean,
        val sampleRate: Double
    )

    private val state = AtomicReference(GateState(enabled = true, sampleRate = 1.0))

    /** Whether new telemetry is currently permitted. Single volatile read. */
    val enabled: Boolean
        get() = state.get().enabled

    /** Current global head-sampling fraction in `[0.0, 1.0]`. Single volatile read. */
    val sampleRate: Double
        get() = state.get().sampleRate

    /** Current immutable snapshot of both control values. */
    fun snapshot(): GateState = state.get()

    /**
     * Applies a parsed [SdkConfig] to the gate.
     *
     * Per the fail-open contract, a `null` [config] (the `sdk` block was absent
     * from an otherwise-valid remote config) resets the gate to the open
     * defaults (`enabled = true, sampleRate = 1.0`). Callers must **not** invoke
     * this on a fetch/parse failure — skipping the call is what preserves the
     * last-applied state across transient errors.
     *
     * @param config the parsed `sdk` block, or `null` when it was absent.
     */
    fun apply(config: SdkConfig?) {
        state.set(
            if (config == null) {
                GateState(enabled = true, sampleRate = 1.0)
            } else {
                GateState(enabled = config.enabled, sampleRate = config.clampedSampleRate)
            }
        )
    }

    /**
     * Hot-path decision for a single event: should it be permitted to flow?
     *
     * Returns `false` when the SDK is disabled, or — when `sampleRate < 1.0` —
     * with probability `1 - sampleRate`, using a non-biased per-thread RNG so
     * the draw allocates nothing and never contends across threads. A
     * `sampleRate` of exactly `1.0` short-circuits the draw entirely.
     *
     * Intended for the log choke point. The span choke point folds [sampleRate]
     * into the sampler's deterministic trace-ID ratio instead (see
     * [io.opentelemetry.android.mobile.sampling.DynamicSampler]).
     *
     * @return `true` to keep the event, `false` to drop it.
     */
    fun allowEvent(): Boolean {
        val current = state.get()
        if (!current.enabled) return false
        val rate = current.sampleRate
        if (rate >= 1.0) return true
        if (rate <= 0.0) return false
        return ThreadLocalRandom.current().nextDouble() < rate
    }
}

File: otel-ios-mobile/Sources/OTelMobileSDK/Policy/RemoteGate.swift
/*
 * Copyright 2025 Barry Solomon
 * SPDX-License-Identifier: Apache-2.0
 */
import Foundation

/// Thread-safe holder for the remote kill-switch + global sampling state
/// `(enabled, sampleRate)` delivered over the wire via the `sdk` block of the
/// DSL v2 config (see `docs/design/remote-kill-switch.md`).
///
/// This is the single source of truth consulted at the two telemetry emission
/// choke points:
///
/// 1. **Logs** — `MobileLogRecordProcessor.onEmit` consults `shouldEmitLog()`
///    *before* any buffering / coalescing work, so a disabled SDK does no work.
/// 2. **Spans** — `RemoteGatedSampler` folds the gate into the trace sampler's
///    `shouldSample` decision.
///
/// `ConfigPoller.applyConfig` pushes parsed `SDKRemoteConfig` here on every
/// successful poll.
///
/// ## Concurrency
///
/// The log choke point (`onEmit`) is **synchronous** per the OTel
/// `LogRecordProcessor` contract, and the span sampler's `shouldSample` is
/// likewise synchronous and on a hot path. Both must read the gate state
/// without `await`-ing — so the gate is a lock-guarded `final class`, **not**
/// an actor. An `os_unfair_lock` guards a single small struct; reads copy the
/// value out under the lock and release immediately. No allocation occurs on
/// the read path, and the per-event sampling RNG is non-biased
/// (`SystemRandomNumberGenerator` via `Double.random(in:)`).
///
/// `os_unfair_lock` is used rather than `NSLock` because the read path is
/// uncontended in the common case (one writer, the poll thread; many readers,
/// the emit threads) and `os_unfair_lock`'s acquire/release is the cheapest
/// primitive available that is still correct under contention.
public final class RemoteGate: @unchecked Sendable {
    private var unfairLock = os_unfair_lock_s()
    private var state: SDKRemoteConfig

    /// Create a gate seeded with `initial` state. Defaults to the
    /// fail-open default (`enabled = true, sampleRate = 1.0`) so a gate that
    /// is never fed config behaves as a no-op — the SDK runs normally.
    public init(initial: SDKRemoteConfig = .default) {
        self.state = initial
    }

    /// Replace the live state. Called by `ConfigPoller.applyConfig` after a
    /// successful parse. Absent `sdk` block ⇒ caller passes `.default`
    /// (re-enable / no restriction); a transient fetch failure never reaches
    /// here, so the last-applied value is preserved automatically.
    public func update(_ config: SDKRemoteConfig) {
        os_unfair_lock_lock(&unfairLock)
        state = config
        os_unfair_lock_unlock(&unfairLock)
    }

    /// Current state snapshot. Cheap copy under the lock.
    public var current: SDKRemoteConfig {
        os_unfair_lock_lock(&unfairLock)
        let snapshot = state
        os_unfair_lock_unlock(&unfairLock)
        return snapshot
    }

    /// Whether the SDK is currently enabled. Convenience read for the
    /// self-telemetry gauge path.
    public var isEnabled: Bool {
        current.enabled
    }

    /// Currently-applied global sample rate, clamped to `[0, 1]`.
    public var sampleRate: Double {
        current.sampleRate
    }

    /// Synchronous, allocation-free decision for the **log** choke point.
    /// Returns `false` (drop) when the SDK is remotely disabled, or
    /// probabilistically when `sampleRate < 1`. Returns `true` (keep)
    /// in normal operation.
    ///
    /// Reads the gate state exactly once, then draws a single non-biased
    /// random sample only when a probabilistic decision is actually required.
    public func shouldEmitLog() -> Bool {
        let snapshot = current
        if !snapshot.enabled { return false }
        return Self.sampleDecision(rate: snapshot.sampleRate)
    }

    /// Folds the gate into a span sampling decision for a span identified by
    /// `traceIdLo` (the low 64 bits of its trace id).
    ///
    /// Semantics (spec §Semantics; matches the Android `DynamicSampler`
    /// MIN/cap fold for cross-platform consistency):
    ///
    /// - `!enabled` ⇒ `.drop` unconditionally — the inner sampler is not
    ///   consulted.
    /// - `sampleRate >= 1.0` (the default / "no restriction") ⇒ `.deferToInner`
    ///   — pure pass-through, zero behavioural change.
    /// - `sampleRate <= 0.0` ⇒ `.drop`.
    /// - otherwise ⇒ a **deterministic trace-id-ratio** test at the global
    ///   rate: `.deferToInner` when `ratio(traceIdLo) < sampleRate`, else
    ///   `.drop`.
    ///
    /// The deterministic ratio is the key to MIN-fold semantics. The inner
    /// sampler (`DynamicSampler` / `traceIdRatio`) applies the *same*
    /// `ratio(traceIdLo) < localRate` threshold test. Two threshold tests on
    /// the one shared ratio compose as intersection — a span survives iff
    /// `ratio < localRate AND ratio < globalRate`, i.e. `ratio < min(local,
    /// global)`. That is exactly `effective = min(localRate, globalRate)`,
    /// and because both decisions key off the same trace id, every span in a
    /// trace gets the same gate verdict (no orphaned children). A random
    /// per-span draw would instead yield the *product* of the rates and break
    /// trace coherence.
    public func spanGate(traceIdLo: UInt64) -> SpanGateDecision {
        let snapshot = current
        if !snapshot.enabled { return .drop }
        let rate = snapshot.sampleRate
        if rate >= 1.0 { return .deferToInner }
        if rate <= 0.0 { return .drop }
        return Self.traceIdRatio(traceIdLo) < rate ? .deferToInner : .drop
    }

    /// Non-biased per-event sampling decision for the **log** path. `rate >= 1`
    /// always keeps, `rate <= 0` always drops, otherwise draw a uniform sample.
    /// Logs have no trace id to key on, so a per-event RNG draw is the correct
    /// (and only) choice here.
    private static func sampleDecision(rate: Double) -> Bool {
        if rate >= 1.0 { return true }
        if rate <= 0.0 { return false }
        return Double.random(in: 0..<1) < rate
    }

    /// Maps the low 64 bits of a trace id to a `[0, 1)` ratio. Mirrors
    /// `DynamicSampler.shouldSampleTraceId` (and Android's `shouldSampleTraceId`)
    /// so the gate's threshold and the inner sampler's threshold operate on the
    /// identical value — the precondition for MIN-fold composition.
    static func traceIdRatio(_ traceIdLo: UInt64) -> Double {
        Double(traceIdLo) / Double(UInt64.max)
    }

    /// Outcome of folding the gate into a span decision.
    public enum SpanGateDecision: Sendable, Equatable {
        /// SDK disabled or globally sampled out — drop the span unconditionally.
        case drop
        /// Gate imposes no restriction — let the inner sampler decide.
        case deferToInner
    }
}
